Specifications/QIR/Quantum-Runtime.md

```diff
@@ -1,47 +1,26 @@
+## Qubits
 
 We define the following functions for managing qubits:
 
 | Function                            | Signature       | Description |
 |-------------------------------------|-----------------|-------------|
 | __quantum__rt__qubit_allocate       | `%Qubit*()`     | Allocates a single qubit. |
-| __quantum__rt__qubit_allocate_array | `%Array*(i64)`  | Allocates an array of qubits. |
-| __quantum__rt__qubit_release        | `void(%Qubit*)` | Release a single qubit. |
-| __quantum__rt__qubit_release_array  | `void(%Array*)` | Release an array of qubits. |
+| __quantum__rt__qubit_allocate_array | `%Array*(i64)`  | Creates an array of the given size and populates it with newly-allocated qubits. |
+| __quantum__rt__qubit_release        | `void(%Qubit*)` | Releases a single qubit. |
+| __quantum__rt__qubit_release_array  | `void(%Array*)` | Releases and array of qubits; each qubit in the array is released, and the array itself is unreferenced. |
 
-Allocated qubits are not guaranteed to be in any particular state.
-If a language guarantees that allocated qubits will be in a specific state, the compiler
-should insert the code required to set the state of the qubits returned from `alloc`.
-Qubits should be unentangled -- measured out -- before they are released.
+The language-specific compiler may assume that qubits are always allocated in a zero-state. Since individual targets may give different guarantees regarding the qubit state upon allocation, the target-specific compilation phase should insert the code required to ensure that the state of qubits is set appropriately. 
 
-If borrowing qubits is supported, then the following runtime functions should also be provided:
-
-| Function                            | Signature       | Description |
-|-------------------------------------|-----------------|-------------|
-| __quantum__rt__qubit_borrow         | `%Qubit*()`     | Borrow a single qubit. |
-| __quantum__rt__qubit_borrow_array   | `%Array*(i64)`  | Borrow an array of qubits. |
-| __quantum__rt__qubit_return         | `void(%Qubit*)` | Return a borrowed qubit. |
-| __quantum__rt__qubit_return_array   | `void(%Array*)` | Return an array of borrowed qubits. |
-
-Borrowing qubits means supplying qubits that are guaranteed not to be otherwise
-accessed while they are borrowed.
-The code that borrows the qubits guarantees that the state of the qubits when
-returned is identical, including entanglement, to their state when borrowed.
-It is always acceptable to satisfy `borrow` by allocating new qubits.
-
-*__Discussion__*
->It will likely be useful to provide usage hints to `alloc` and `borrow`.
->Since we don't know yet what form these hints may take, we leave defining them
->for future work.
+Any measurements or resets applied upon release are at the discretion of the target as well; the qubits that are to be released hence need to be unentangled from qubits that remain live prior to invoking the release function.
```

Comment:
I might reword the second phrase a bit here. There's no QIR requirement that the qubits be unentangled; we won't check that, and we won't generate an error if they are. Rather, behavior of the quantum algorithm might be unpredictable if qubits are released that are still entangled with active program qubits.

Reply:
Rephrased.